text/0000-tangle/0000-tangle.md

```diff
@@ -0,0 +1,409 @@
++ Feature name: tangle
++ Start date: 2019-10-22
++ RFC PR: [iotaledger/bee-rfcs#0000](https://github.com/iotaledger/bee-rfcs/pull/0000)
++ Bee issue: [iotaledger/bee#0000](https://github.com/iotaledger/bee/issues/0000)
+
+# Summary
+
+The Tangle is the core of IOTA's distributed ledger technology. This RFC proposes an API for interacting with and manipulating the Tangle in a type-safe manner that mirrors the flexible semantics of a garbage-collected language with retaining the performance and ownership model of Rust. It also proposes a novel eager solidification algorithm that is (likely to be) more performant than the existing thread-based solidification system, as well as only ever providing users of the API with a fully-informed, fully-inferred view of Tangle solidification.
+
+# Motivation
+
+## API
+
+Rust's ownership model is designed to prevent concurrent mutable aliasing of a value. It does this by tracking the lifetimes of mutable references and enforcing an ownership model that permits each value to be owned by only one other thing.
+
+The Tangle, being a recursive data structure that permits ownership relationships that do not pertain to Rust's ownership model, cannot be implemented using Rust's ownership semantics and must instead be modelled using a data structure such as a `HashMap`. As a result, graph APIs written in Rust tend to be rather more awkward to manipulate than those of garbage-collected or memory-unsafe languages.
+
+For example, in a traditional `HashMap`-driven API:
+
+```rust
+let mut tangle = Tangle::default();
+
+// Insert a transaction
+let tx0_hash = tangle.insert(tx0);
+
+// Navigate from tx0 to the trunk of the trunk of tx0
+let tx0_trunk_hash = tangle
+    .get(tx0_hash)
+    .unwrap()
+    .trunk;
+let tx0_trunk_trunk_hash = tangle
+    .get(tx0_trunk_hash)
+    .unwrap()
+    .trunk;
+let tx0_trunk_trunk = tangle
+    .get(tx0_trunk_trunk_hash)
+    .unwrap();
+```
+
+As can be seen, walking the Tangle is unnecessarily explicit.
+
+To address this issue we suggest a design that makes use of Rust's zero-cost approach to abstraction to transparently manipulate the Tangle using guard/reference types and the `Deref` trait.
```

Comment:
I think it's also worth mentioning `Deref` implicitly implements all the methods of its associate type.